### ISSUE TITLE
SSIP Description of Progress-Prioriy Messages

### ISSUE THREAD

**Issue by BoniLindsley** [brailcom/speechd#593]
I am reading the [SSIP specification][], and I would like some clarifications on how it thinks messages of priority `progress` received by a speech server should be handled. My guess is that it works as follows: when a `progress` message is received by a speech server,

-   If there are mssages with priorities `important`, `message` or `text` waiting in the queues or being spoken, the incoming message is cancelled.
-   From here, we assume the messages in the queue or being spoken, if any, are of priorities `progress` or `notification`.
-   If there are `progress` messages in the queue, queued messages are cancelled, and the incoming message is queued.
-   If there are no `progress` messages in the queue but the message being spoken has `progress` priority, the incoming message is queued.
-   Lastly, note that `notification` messages do not get queued, and when they are being spoken, there are no queued messages. This means there is only one remaining situation to handle: there is a `notification` message being spoken, and the queue is empty. In such a situation, the `notification` message is cancelled, and the incoming `progress` message is spoken.

I would like to know whether this is correct or not. And if not, please point out where my mistakes are.

The remainder of this post describes my thought process in arriving at the above guess. So feel free to ignore it if you already have enough information to answer.

### A Description of my Understanding

This is a little long, since I need to explain the logic that went through my mind. It centres around the part concerning detection of a last `progress` message.

An incoming message of priority `progress` behaves as priority `notification` (with two caveats that will be expanded on below), according to <https://github.com/brailcom/speechd/blob/80bf1902f471d7abf227a1be15e0a5c6c9e0b7e2/doc/ssip.texi#L637-L651>

Behaving as a `notification` means the incoming message is cancelled if there are any pending messages (either speaking or queued) that are not of priority `notification`, and cancels the speaking message to speak if it is of `notification` priority, according to <https://github.com/brailcom/speechd/blob/80bf1902f471d7abf227a1be15e0a5c6c9e0b7e2/doc/ssip.texi#L618-L625>

This is consistent with the first caveat which says that the incoming message cancels itself rather than the speaking message if the speaking message is of `progress` priority.

The information so far suggests the following: One, the incoming `progress` message is cancelled if there are any pending messages not of priority `notification`. Two: the incoming message is spoken otherwise, by cancelling all pending `notification` messages if necessary. In particular, much like a `notification` message, `progress` messages are not queued -- they are either cancelled by priority or spoken immediately by cancelling something else.

### Problem

My confusion then comes from the wording of the second caveat. It says that when "there are no more messages of this priority waiting in the queues and if the last of them wasn't spoken yet, it speaks it with the priority `message`". But I had concluded, perhaps falsely, that `progress` messages are not queued, leading to a contradiction in my mind. This probably comes from a mix-up of terminology or a misunderstanding. It would be great if someone can point out where my mistakes are

Now, the goal of the second caveat, as it mentions and as I understand it, is to allow final `progress` messages to be spoken. This necessarily means that not all unspoken `progress` messages are cancelled. Working backwards from this, is how I had arrived in the guess at the beginning of this post.

[SSIP specification]: https://github.com/brailcom/speechd/blob/80bf1902f471d7abf227a1be15e0a5c6c9e0b7e2/doc/ssip.texi

**Comment by sthibaul** [comment]
Hello,

I'm sorry it took long to get an answer. To tell the truth since I'm not the original author of speech-dispatcher I don't know what SSIP says about this and what the code does (I didn't even know about `notification` and `progress` priorities), so I had to take time to understand what you wrote. What you are saying seems about right to me. The only thing I found perhaps a bit odd is that a progress message would interrupt a notification message. I would have thought we'd want to let the notification complete before saying the progression.

> to allow final progress messages to be spoken. This necessarily means that not all unspoken progress messages are cancelled

Yes, people don't care about getting all of the progression details, they only need to know about the latest. But they do want to know about the latest.

I would say experimentation would show what the behavior of the current code is, and possibly rephrase the documentation accordingly. Because the current code was used by brailcom for their own users, so the current behavior is probably what is wanted, even if the documentation may let its reader understand something else.

**Comment by BoniLindsley** [comment]
Thank you very much for answering. Good to know I can still do a little bit of logical thinking!

I also found notifications being interrupted a little weird. But that was in the specification... There are a few other things unclear in the specification.

But well... I guess, from here, I should treat the specification more as a starting point and use speech dispatcher itself as a reference implementation, like how CPython is to Python. It's going to be a lot of testing!

Thanks again for reading through my question! I know it was quite a wall of text. I had already tried to be as succint as I could in phrasing and describing the logic!
